### ISSUE TITLE
Help with swagger configuration on localhost (typescript)

### ISSUE THREAD

**Issue by weidemannadrien** [fastify/help#383]
#### What are you trying to achieve, or the steps to reproduce?

Make Swagger working in local developpment: localhost:PORT/API-VER/api-docs
Example: localhost:4000/v1/api-docs

My fastify initialization:
```js
//app.ts
import { Server, IncomingMessage, ServerResponse } from "http";
import path from "path";
import {
  fastify,
  FastifyInstance,
  FastifyRequest,
  FastifyReply
} from "fastify";
import autoLoad from "fastify-autoload";
import cors from "fastify-cors";
import helmet from "fastify-helmet";
import fastifySwagger from "fastify-swagger";
import fastifyFormbody from "fastify-formbody";
import pino from "pino";
import HttpStatus from "http-status-codes";

import ENV from "./env";

import { GLOBAL } from "./constants";

import swaggerConfig from "./swagger-base.config.json";

class App {
  public readonly server: FastifyInstance<
    Server,
    IncomingMessage,
    ServerResponse
  >;
  private readonly port: number;

  constructor(port: number) {
    const logger = pino({
      enabled: ENV.IS_LOGGER_ENABLED,
      level: ENV.LOGGER_LEVEL_CONSOLE,
      formatters: {
        level: (label: string): { level: string } => {
          return { level: label };
        }
      }
    });

    this.server = fastify({
      logger,
      bodyLimit: GLOBAL.BODY_LIMIT
    });

    this.port = port;

    void this.initSwagger();
    void this.initRoutes();
    void this.initMiddleware();
    this.initErrorMiddleware();
  }

  public async listen(): Promise<void> {
    await this.server.listen(Number(process.env.PORT), process.env.HOST);
  }

  private async initRoutes(): Promise<void> {
    const API_VERSION: string = ENV.API_VERSION as string;

    await this.server.register(autoLoad, {
      dir: path.join(__dirname, "routes"),
      options: { prefix: `/${API_VERSION}` }
    });

    // 404 handler (ALWAYS keep this as the last route)
    this.server.all("/*", (request: FastifyRequest, reply: FastifyReply) => {
      const url: string = request.url;
      const method: string = request.method;

      /*const error = new CustomError({
        name: ERROR.NAME.ROUTER,
        message: `Route ${url} not found with ${method} method`,
        status: HttpStatus.NOT_FOUND
      });*/

      void reply.status(HttpStatus.NOT_FOUND).send({ error: "Error" });
    });
  }

  private async initSwagger(): Promise<void> {
    //only want my swagger docs in local/development env
    if (
      ENV.NODE_ENV !== undefined &&
      (ENV.NODE_ENV === GLOBAL.ENV.LOCAL ||
        ENV.NODE_ENV === GLOBAL.ENV.DEVELOPMENT)
    ) {
      const API_VERSION: string = ENV.API_VERSION as string;

      await this.server.register(fastifySwagger, {
        routePrefix: `${API_VERSION}/api-docs`,
        swagger: swaggerConfig.definition,
        exposeRoute: true
      });
    }
  }

  private async initMiddleware(): Promise<void> {
    await this.server.register(cors, {
      methods: ["HEAD", "GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"],
      allowedHeaders: [
        "Content-Type",
        "Authorization",
        GLOBAL.API_KEY.HEADER_KEY_NAME
      ]
    });

    await this.server.register(fastifyFormbody);


    // tried solution thanks to the help of the Discord community (thanks!)
    await this.server.register(helmet, {
      contentSecurityPolicy: {
        directives: {
          defaultSrc: ["'self'"],
          baseUri: ["'self'"],
          blockAllMixedContent: [],
          fontSrc: ["'self'", "https:", "data:"],
          frameAncestors: ["'self'"],
          imgSrc: ["'self'", "data:"],
          objectSrc: ["'none'"],
          scriptSrc: ["'self'"].concat(this.server.swaggerCSP.script),
          scriptSrcAttr: ["'none'"],
          styleSrc: ["'self'", "https:"].concat(this.server.swaggerCSP.style),
          upgradeInsecureRequests: []
        }
      }
    });

    this.server.addHook("onRequest", maintenanceMiddleware);
  }

  private initErrorMiddleware(): void {
    this.server.setErrorHandler(errorMiddleware);
  }
}

export default App;
```

// swagger-base.config.json
```json
{
  "definition": {
    "openapi": "3.0.2",
    "servers": [
      {
        "url": "http://localhost:4040/v1",
        "description": "Local server instance"
      },
      {
        "url": "https://sample2.mybusiness.com",
        "description": "Development server instance"
      },
      {
        "url": "https://sample.mybusiness.com",
        "description": "Production server instance"
      }
    ],
    "host": "localhost",
    "externalDocs": {
      "url": "https://mybusiness.com/sample-api",
      "description": "mybusiness link"
    },
    "info": {
      "title": "SAMPLE API",
      "description": "This Swagger design sample api routes.",
      "version": "1.0.0",
      "contact": {
        "email": "mybusiness@mybusiness.com"
      },
      "license": {
        "name": "Private"
      }
    },
    "components": {
      "securitySchemes": {
        "bearerAuth": {
          "type": "http",
          "scheme": "bearer",
          "bearerFormat": "JWT"
        }
      }
    }
  }
}
```

#### What was the result you received?
Redirected on http://localhost:4000/v1/api-docs/static/index.html with a blank page and give me this error in console: 
Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self'". Either the 'unsafe-inline' keyword, a hash ('sha256-....'), or a nonce ('nonce-...') is required to enable inline execution.

#### What did you expect?
My beautiful swagger docs

#### Context

* *Chrome*: Version 88.0.4324.192 (Official Build) (x86_64)
* *node version*: 14
* *fastify*: "^3.13.0
* *fastify-autoload*: ^3.5.2
* *fastify-cors*: ^5.2.0
* *fastify-formbody*: ^5.0.0
* *fastify-helmet*: ^5.2.0
* *fastify-swagger*: ^4.3.3
* *swagger-schema-official*: ^2.0.0-bab6bed
* *typescript*: ^4.2.2

**Comment by weidemannadrien** [comment]
Thanks for you reply.

I edited my code as you suggest, that doesn't work either. I don't understand the full logic of Helmet params, that's my problem. But I took your earlier response, so I guess the problem is elsewhere.

**Comment by climba03003** [comment]
> 
> 
> Thanks for you reply.
> 
> I edited my code as you suggest, that doesn't work either. I don't understand the full logic of Helmet params, that's my problem. But I took your earlier response, so I guess the problem is elsewhere.

Can you describe more about your issue? Such as the error message shown on console, browser, etc.

**Comment by weidemannadrien** [comment]
Hello, I noticed something.

I've no error print in my console, so I didn't think about it, but when I tried to log instance.swaggerCSP, I think I have an error. My console print nothing:

```
private async initMiddleware(): Promise<void> {
    await this.server.register(cors, {
      methods: ["HEAD", "GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"],
      allowedHeaders: [
        "Content-Type",
        "Authorization",
        GLOBAL.API_KEY.HEADER_KEY_NAME
      ]
    });

    await this.server.register(fastifyFormbody);

    console.log("hello world");
    console.log(this.server.swaggerCSP.script);
    console.log("hello world 2");
    await this.server.register(helmet, {
      contentSecurityPolicy: {
        directives: {
          defaultSrc: ["'self'"],
          baseUri: ["'self'"],
          blockAllMixedContent: [],
          fontSrc: ["'self'", "https:", "data:"],
          frameAncestors: ["'self'"],
          imgSrc: ["'self'", "data:"],
          objectSrc: ["'none'"],
          scriptSrc: ["'self'"].concat(this.server.swaggerCSP.script),
          scriptSrcAttr: ["'none'"],
          styleSrc: ["'self'", "https:"].concat(this.server.swaggerCSP.style),
          upgradeInsecureRequests: []
        }
      }
    });

    this.server.addHook("onRequest", maintenanceMiddleware);
  }
```

And there is my result:
```
9:29:11 AM - Found 0 errors. Watching for file changes.
hello world
```
The code break before "hello world 2". Any idea?

**Comment by simoneb** [comment]
See how we're doing it here https://github.com/nearform/titus/blob/master/packages/titus-backend/lib/server.js

**Comment by weidemannadrien** [comment]
Ok, thanks to your help guys @simoneb and @climba03003, that works now. I don't know exactly how I managed to resolve my issue, but I put my endpoint autoload at the top of my init function, then swagger init, then my middleware init (including helmet). And I placed Helmet before CORS or whatever, at the top of the initMiddleware() function.

```js
async init() {
    await this.initRoutes();
    await this.initSwagger();
    await this.initMiddleware();
    this.initErrorMiddleware();
}
```

I put this issue as resolve. Have a nice day!
